src/m3ninx/index/segment/fst/encoding/docs/data.go

```diff
@@ -128,3 +129,112 @@ func (r *DataReader) Read(offset uint64) (doc.Metadata, error) {
 
 	return d, nil
 }
+
+// EncodedDataReader is a reader for the data file for encoded document metadata.
+type EncodedDataReader struct {
+	data []byte
+}
+
+// NewEncodedDataReader returns a new EncodedDataReader.
+func NewEncodedDataReader(data []byte) *EncodedDataReader {
+	return &EncodedDataReader{
+		data: data,
+	}
+}
+
+// Read reads a doc.Encoded from a data stream starting at the specified offset.
+func (e *EncodedDataReader) Read(offset uint64) (doc.Encoded, error) {
+	if offset >= uint64(len(e.data)) {
+		return doc.Encoded{}, fmt.Errorf(
+			"invalid offset: %v is past the end of the data file", offset,
+		)
+	}
+
+	return doc.Encoded{
+		Bytes: e.data[int(offset):],
+	}, nil
+}
+
+// EncodedDocumentReader is a reader for reading documents from encoded metadata.
+type EncodedDocumentReader struct {
+	currFields []doc.Field
+}
+
+// NewEncodedDocumentReader returns a new EncodedDocumentReader.
+func NewEncodedDocumentReader() *EncodedDocumentReader {
+	return &EncodedDocumentReader{}
+}
+
+// Read reads a doc.Metadata from a doc.Encoded. Returned doc.Metadata should be
+// processed before calling Read again as the underlying array pointed to by the Fields
+// slice will be updated. This approach avoids allocating a new slice with a new backing
+// array for every document processed, unlike (*DataReader).Read
+func (r *EncodedDocumentReader) Read(encoded doc.Encoded) (doc.Metadata, error) {
+	for i := range r.currFields {
+		r.currFields[i] = doc.Field{}
+	}
+	r.currFields = r.currFields[:0]
+	id, buf, err := encoding.ReadBytes(encoded.Bytes)
+	if err != nil {
+		return doc.Metadata{}, err
+	}
+
+	x, buf, err := encoding.ReadUvarint(buf)
+	if err != nil {
+		return doc.Metadata{}, err
+	}
+	n := int(x)
+
+	var name, val []byte
+	for i := 0; i < n; i++ {
+		name, buf, err = encoding.ReadBytes(buf)
+		if err != nil {
+			return doc.Metadata{}, err
+		}
+		val, buf, err = encoding.ReadBytes(buf)
+		if err != nil {
+			return doc.Metadata{}, err
+		}
+		r.currFields = append(r.currFields, doc.Field{
+			Name:  name,
+			Value: val,
+		})
+	}
+
+	return doc.Metadata{
+		ID:     id,
+		Fields: r.currFields,
+	}, nil
+}
+
+// ReadEncodedDocumentID reads the document ID from the encoded document metadata.
+func ReadEncodedDocumentID(encoded doc.Encoded) ([]byte, error) {
+	id, _, err := encoding.ReadBytes(encoded.Bytes)
+	return id, err
+}
+
+// GetFromDocument retrieves a doc.Metadata from a doc.Document.
+func GetFromDocument(document doc.Document, reader *EncodedDocumentReader) (doc.Metadata, error) {
```

Comment:
nit: Go naming convention tries to downplay using word "Get" as much as possible (https://golang.org/doc/effective_go.html#Getters) so perhaps this could just be `MetadataFromDocument`?

Reply:
Maybe `MetadataFromDocument`

Reply:
Good point. Will fix this in #3057 so as to not wait on the build again.